### PR TITLE
[SYCL] Refactor address space casts functionality

### PR DESCRIPTION
**Pull request by aelovikov-intel**
This follows the interfaces designed in https://github.com/intel/llvm/blob/3a1c3cb53566f904a73361d5c57b939d981564b5/sycl/doc/extensions/proposed/sycl_ext_oneapi_address_cast.asciidoc, but instead of operating on `multi_ptr`, these work on decorated C++ pointers (as that's what we need throughout our implementation, including `multi_ptr` implementation itself).

Basically, I've moved the implementation of the extension to the new `detail::static|dynamic_address_cast` functions and replaced all uses of the old `detail::cast_AS` (that had inconsistent static vs dynamic behavior depending on address spaces/backends) and also uses of direct SPIRV builtin/wrappers invocations.

This isn't NFC, because by doing that I've changed "dynamic" behavior to "static" whenever the spec allows that (e.g. if it's UB if runtime pointers doesn't point to a proper allocation).

### CODE REVIEW

File: sycl/include/sycl/ext/oneapi/experimental/prefetch.hpp [hunk 1]
```diff
@@ -71,7 +71,10 @@ inline constexpr bool check_prefetch_acc_mode =
 template <typename T, typename Properties>
 void prefetch_impl(T *ptr, size_t bytes, Properties properties) {
 #ifdef __SYCL_DEVICE_ONLY__
-  auto *ptrGlobalAS = __SYCL_GenericCastToPtrExplicit_ToGlobal<const char>(ptr);
+  auto *ptrGlobalAS =
+      reinterpret_cast<__attribute__((opencl_global)) const char *>(
+          detail::static_address_cast<access::address_space::global_space>(
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/experimental/prefetch.hpp:76]: See https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_prefetch.asciidoc, it states (for non-accessor overloads):
> Preconditions: ptr must point to an object in global memory.

File: sycl/include/sycl/group.hpp [hunk 2]
```diff
@@ -449,12 +449,12 @@ template <int Dimensions = 1> class __SYCL_TYPE(group) group {
     using QualSrcT =
         std::conditional_t<std::is_const_v<SrcT>, const uint8_t, uint8_t>;
     auto DestP = multi_ptr<uint8_t, DestS, access::decorated::yes>(
-        detail::cast_AS<typename multi_ptr<uint8_t, DestS,
-                                           access::decorated::yes>::pointer>(
+        reinterpret_cast<typename multi_ptr<uint8_t, DestS,
```
Comment (aelovikov-intel) [inline sycl/include/sycl/group.hpp:452]: We're changing element type of the decorated C++ pointer type here, no address space changes. Same throughout this file.

File: sycl/include/sycl/nd_item.hpp [hunk 3]
```diff
@@ -384,12 +384,12 @@ template <int Dimensions = 1> class nd_item {
     using QualSrcT =
         std::conditional_t<std::is_const_v<SrcT>, const uint8_t, uint8_t>;
     auto DestP = multi_ptr<uint8_t, DestS, access::decorated::yes>(
-        detail::cast_AS<typename multi_ptr<uint8_t, DestS,
-                                           access::decorated::yes>::pointer>(
+        reinterpret_cast<typename multi_ptr<uint8_t, DestS,
```
Comment (aelovikov-intel) [inline sycl/include/sycl/nd_item.hpp:387]: We're changing element type of the decorated C++ pointer type here, no address space changes. Same throughout this file.

File: sycl/include/sycl/multi_ptr.hpp [hunk 4]
```diff
@@ -272,7 +272,7 @@ class __SYCL_TYPE(multi_ptr) multi_ptr {
                            OtherSpace != access::address_space::constant_space>>
   multi_ptr &
   operator=(const multi_ptr<value_type, OtherSpace, OtherIsDecorated> &Other) {
-    m_Pointer = detail::cast_AS<decorated_type *>(Other.get_decorated());
+    m_Pointer = detail::static_address_cast<Space>(Other.get_decorated());
```
Comment (aelovikov-intel) [inline sycl/include/sycl/multi_ptr.hpp:275]: "static" cast from non-`constant_space` to `generic_space` is always possible.

File: sycl/include/sycl/multi_ptr.hpp [hunk 5]
```diff
@@ -282,15 +282,18 @@ class __SYCL_TYPE(multi_ptr) multi_ptr {
                            OtherSpace != access::address_space::constant_space>>
   multi_ptr &
   operator=(multi_ptr<value_type, OtherSpace, OtherIsDecorated> &&Other) {
-    m_Pointer = detail::cast_AS<decorated_type *>(std::move(Other.m_Pointer));
+    m_Pointer = detail::static_address_cast<Space>(std::move(Other.m_Pointer));
```
Comment (aelovikov-intel) [inline sycl/include/sycl/multi_ptr.hpp:285]: Likewise.

File: sycl/include/sycl/multi_ptr.hpp [hunk 6]
```diff
@@ -282,15 +282,18 @@ class __SYCL_TYPE(multi_ptr) multi_ptr {
...
   }
 
   reference operator*() const { return *m_Pointer; }
   pointer operator->() const { return get(); }
   reference operator[](difference_type index) const { return m_Pointer[index]; }
 
-  pointer get() const { return detail::cast_AS<pointer>(m_Pointer); }
+  pointer get() const {
+    return detail::static_address_cast<
+        is_decorated ? Space : access::address_space::generic_space>(m_Pointer);
```
Comment (aelovikov-intel) [inline sycl/include/sycl/multi_ptr.hpp:295]: Either no-op/identity or cast away decoration, always possible.

File: sycl/include/sycl/multi_ptr.hpp [hunk 7]
```diff
@@ -312,9 +315,7 @@ class __SYCL_TYPE(multi_ptr) multi_ptr {
   explicit
   operator multi_ptr<value_type, OtherSpace, OtherIsDecorated>() const {
     return multi_ptr<value_type, OtherSpace, OtherIsDecorated>{
-        detail::cast_AS<typename multi_ptr<value_type, OtherSpace,
-                                           access::decorated::yes>::pointer>(
-            get_decorated())};
+        detail::static_address_cast<OtherSpace>(get_decorated())};
```
Comment (aelovikov-intel) [inline sycl/include/sycl/multi_ptr.hpp:318]: Conversions from `generic_ptr` are defined such that it's UB if the pointer doesn't address the destination adddres space, same for cases below.

File: sycl/include/sycl/multi_ptr.hpp [hunk 8]
```diff
@@ -353,17 +352,14 @@ class __SYCL_TYPE(multi_ptr) multi_ptr {
...
         get_decorated())};
   }
 
   template <access::decorated ConvIsDecorated>
   operator multi_ptr<const value_type, Space, ConvIsDecorated>() const {
-    return multi_ptr<const value_type, Space, ConvIsDecorated>{
-        detail::cast_AS<typename multi_ptr<const value_type, Space,
-                                           access::decorated::yes>::pointer>(
-            get_decorated())};
+    return multi_ptr<const value_type, Space, ConvIsDecorated>{get_decorated()};
```
Comment (aelovikov-intel) [inline sycl/include/sycl/multi_ptr.hpp:362]: Regardless of the `access::decoration`, internal raw pointer is always decorated and that's what ctor accepts, no address cast is happening here.

File: sycl/include/sycl/multi_ptr.hpp [hunk 9]
```diff
@@ -386,11 +382,8 @@ class __SYCL_TYPE(multi_ptr) multi_ptr {
            Space == access::address_space::ext_intel_global_host_space)>>
   explicit
   operator multi_ptr<ElementType, GlobalSpace, DecorateAddress>() const {
-    using global_pointer_t =
-        typename multi_ptr<ElementType, GlobalSpace,
-                           access::decorated::yes>::pointer;
     return multi_ptr<ElementType, GlobalSpace, DecorateAddress>(
-        detail::cast_AS<global_pointer_t>(get_decorated()));
+        detail::static_address_cast<GlobalSpace>(get_decorated()));
```
Comment (aelovikov-intel) [inline sycl/include/sycl/multi_ptr.hpp:386]: These extension address spaces *are* subsections of the `global_space`, so the cast is known to succeed.

File: sycl/include/sycl/multi_ptr.hpp [hunk 10]
```diff
@@ -487,7 +480,7 @@ class __SYCL_TYPE(multi_ptr) multi_ptr<const void, Space, DecorateAddress> {
   multi_ptr(accessor<ElementType, Dimensions, Mode, target::device,
                      isPlaceholder, PropertyListT>
                 Accessor)
-      : multi_ptr(detail::cast_AS<decorated_type *>(
+      : multi_ptr(detail::static_address_cast<Space>(
```
Comment (aelovikov-intel) [inline sycl/include/sycl/multi_ptr.hpp:483]: `target::device` accessor does point to the `global_space` (or `ext_intel_global_device_space` when the extension is used).

File: sycl/include/sycl/multi_ptr.hpp [hunk 11]
```diff
@@ -552,10 +548,9 @@ class __SYCL_TYPE(multi_ptr) multi_ptr<const void, Space, DecorateAddress> {
   template <typename ElementType,
             typename = typename std::enable_if_t<std::is_const_v<ElementType>>>
   explicit operator multi_ptr<ElementType, Space, DecorateAddress>() const {
-    return multi_ptr<ElementType, Space, DecorateAddress>{
-        detail::cast_AS<typename multi_ptr<ElementType, Space,
+    return {static_cast<typename multi_ptr<ElementType, Space,
```
Comment (aelovikov-intel) [inline sycl/include/sycl/multi_ptr.hpp:551]: We are casting from a `__decorated__ const void *` to `__decorated__ const something *` here.

File: sycl/include/sycl/multi_ptr.hpp [hunk 12]
```diff
@@ -579,11 +574,8 @@ class __SYCL_TYPE(multi_ptr) multi_ptr<const void, Space, DecorateAddress> {
            Space == access::address_space::ext_intel_global_host_space)>>
   explicit
   operator multi_ptr<const void, GlobalSpace, DecorateAddress>() const {
-    using global_pointer_t =
-        typename multi_ptr<const void, GlobalSpace,
-                           access::decorated::yes>::pointer;
-    return multi_ptr<const void, GlobalSpace, DecorateAddress>(
-        detail::cast_AS<global_pointer_t>(m_Pointer));
+    return multi_ptr<const void, GlobalSpace, DecorateAddress>{
+        detail::static_address_cast<GlobalSpace>(m_Pointer)};
```
Comment (aelovikov-intel) [inline sycl/include/sycl/multi_ptr.hpp:578]: These address spaces are part of the `global_space`, always succeeds.

File: sycl/include/sycl/multi_ptr.hpp [hunk 13]
```diff
@@ -638,7 +630,7 @@ class __SYCL_TYPE(multi_ptr) multi_ptr<void, Space, DecorateAddress> {
   multi_ptr(accessor<ElementType, Dimensions, Mode, target::device,
                      isPlaceholder, PropertyListT>
                 Accessor)
-      : multi_ptr(detail::cast_AS<decorated_type *>(
+      : multi_ptr(detail::static_address_cast<Space>(
```
Comment (aelovikov-intel) [inline sycl/include/sycl/multi_ptr.hpp:633]: because accessor.

File: sycl/include/sycl/multi_ptr.hpp [hunk 14]
```diff
@@ -692,7 +684,10 @@ class __SYCL_TYPE(multi_ptr) multi_ptr<void, Space, DecorateAddress> {
     return *this;
   }
 
-  pointer get() const { return detail::cast_AS<pointer>(m_Pointer); }
+  pointer get() const {
+    return detail::static_address_cast<
+        is_decorated ? Space : access::address_space::generic_space>(m_Pointer);
```
Comment (aelovikov-intel) [inline sycl/include/sycl/multi_ptr.hpp:689]: See above - either no-op/identity or cast away decoration.

File: sycl/include/sycl/multi_ptr.hpp [hunk 15]
```diff
@@ -703,8 +698,8 @@ class __SYCL_TYPE(multi_ptr) multi_ptr<void, Space, DecorateAddress> {
   template <typename ElementType>
   explicit operator multi_ptr<ElementType, Space, DecorateAddress>() const {
     return multi_ptr<ElementType, Space, DecorateAddress>{
-        detail::cast_AS<typename multi_ptr<ElementType, Space,
-                                           access::decorated::yes>::pointer>(
+        static_cast<typename multi_ptr<ElementType, Space,
+                                       access::decorated::yes>::pointer>(
```
Comment (aelovikov-intel) [inline sycl/include/sycl/multi_ptr.hpp:702]: Changing element type from `void` to `ElementType`, no address space changes.

File: sycl/include/sycl/multi_ptr.hpp [hunk 16]
```diff
@@ -728,10 +723,8 @@ class __SYCL_TYPE(multi_ptr) multi_ptr<void, Space, DecorateAddress> {
           (Space == access::address_space::ext_intel_global_device_space ||
            Space == access::address_space::ext_intel_global_host_space)>>
   explicit operator multi_ptr<void, GlobalSpace, DecorateAddress>() const {
-    using global_pointer_t =
-        typename multi_ptr<void, GlobalSpace, access::decorated::yes>::pointer;
     return multi_ptr<void, GlobalSpace, DecorateAddress>(
-        detail::cast_AS<global_pointer_t>(m_Pointer));
+        detail::static_address_cast<GlobalSpace>(m_Pointer));
```
Comment (aelovikov-intel) [inline sycl/include/sycl/multi_ptr.hpp:727]: Subsections of the `global_space` per definition, known to succeed.

File: sycl/include/sycl/multi_ptr.hpp [hunk 17]
```diff
@@ -782,14 +775,17 @@ class __SYCL2020_DEPRECATED(
 #endif
 
   multi_ptr(ElementType *pointer)
-      : m_Pointer(detail::cast_AS<pointer_t>(pointer)) {
+      : m_Pointer(detail::dynamic_address_cast<
```
Comment (aelovikov-intel) [inline sycl/include/sycl/multi_ptr.hpp:778]: See comment near the definition of `detail::dynamic_address_cast`, that part is mostly NFC. The exception is non-SPIRV backends when we know that address spaces are *NOT* compatible. Previously we were doing C-style cast anyway, now `nullptr` is returned.

Same for ctors below.

File: sycl/include/sycl/ext/oneapi/experimental/address_cast.hpp [hunk 18]
```diff
@@ -63,39 +41,14 @@ multi_ptr<ElementType, Space, DecorateAddress> static_address_cast(
   if constexpr (Space == generic_space)
     return Ptr;
   else
-    return {static_address_cast<Space>(Ptr.get_raw())};
+    return {static_address_cast<Space>(Ptr.get_decorated())};
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/experimental/address_cast.hpp:44]: We were limited by the extension interface before, so had to drop this bit of information in order to delegate. Now that we're calling something from `detail::` we can keep decoration in place.

**Review by bso-intel — APPROVED** [review]
LGTM.
Thanks for extra comments about your code changes.

**Review by steffenlarsen — APPROVED** [review]
Very nice!

**Pull request merged**